### ISSUE TITLE
spec-char-escape will bite angular apps

### ISSUE THREAD

**Issue by KrekkieD** [htmllint/htmllint#123]
The rule will not allow angular logic in attributes, i.e. `<p ng-hide="yes && sure"></p>`.

I was working on a regex-exclude-rule (much like the `id-class-style` rule), but figured this would only provide a partial solution. The regex I was working on would exclude attributes starting with `ng-`.

The issue is that custom directives will also allow logic in the attribute values (`<p my-custom-error="valid && notEmpty"></p>`), these do not use the `ng-` naming convention and would therefore need to be excluded manually, which could become a pain.

Another possibility might be to set an `environment` option, as f.e. `jshint` does (`"angular": "true"`). This could enable integrated attribute exclusion for certain rules. However, this would still not prevent the custom directive issues.

Perhaps a better solution can be found for this.

**Comment by KrekkieD** [comment]
Is this project still alive?

**Comment by mlochbaum** [comment]
It's still alive, but the owners are all a bit unmotivated as we don't use the linter. Personally I am trying to finish a master's thesis this semester.

I'm not really sure why spec-char-escape checks tags: a tag name with a strange symbol in it is usually poor style but shouldn't introduce any unexpected behavior like it would in text. So the correct solution may just be to remove 'tag' from the filter property of that rule.

**Comment by KrekkieD** [comment]
The tag itself (i.e. `ng-hide`) is fine, but it breaks on the `&&` part of the tag value (`"yes && sure"`). This sort of logic is not uncommon for angular applications and could include many javascript-like expressions.

**Comment by mlochbaum** [comment]
Sorry, I meant "tag value" rather than "tag name". Since most attribute values aren't displayed by the browser as text, they don't need to be escaped (unless I've misunderstood something about html). However, a few types are, such as `alt` for images or `value` for input fields. Ideally htmllint should restrict the rule to these kinds of attributes. I'm having trouble finding a list of the attributes whose values will be displayed as text, though.

**Comment by nkoterba** [comment]
Just chiming in here...

In our angular app, we have HTML like:
```html
<md-button ng-disabled="$ctrl.editWorkspaceForm.$invalid ||
                 $ctrl.missingAdmin"
</md-button>
```

Unfortunately, `|` is considered an "unsafe" character although it's valid and ok in this syntax.  Is there a way to customize the unsafe or invalid characters checked?

What might be even better is some type of regex where I can say ignore these elements for the attr-no-unsafe-char rule.

Thanks!

**Comment by mlochbaum** [comment]
`|` isn't treated as unsafe--did you mean `&`? And I hope the actual html has an extra closing bracket... In any case, the `text-ignore-regex` option in v0.7.0 does what you want. Set it to '&' to ignore ampersands, or maybe '[^]*$[^]*' to ignore anything with a `$` in it.

**Comment by mlochbaum** [comment]
Looks like `<&"` in attribute values should be escaped (in plain html) regardless of how the attribute value will be used, so I'll retract my earlier comment about checking only values that will be displayed as text. `text-ignore-regex` addresses this issue well enough, I think.
